src/widgets/skills.ts

```diff
@@ -1,29 +1,126 @@
 import buildGradientBox from '../components/gradient-box'
 import { findData } from '../utils'
-import errorWidget from './error'
 import languageData from '../data/languages'
+import frameworks from '../data/frameworks'
+import libraries from '../data/libraries'
+import tools from '../data/tools'
+import softwareIDEs from '../data/software-ides'
 
+/**
+ * Builds the skill widget page
+ * 
+ * This method builds a SVG file that contains all the languages, frameworks,
+ * libraries, tools and other software that can be found in the data folder and
+ * are passed by the calling method.
+ * 
+ * @param languagesString The string with all the languages
+ * @param frameworksString The string with all the frameworks
+ * @param librariesString The string with all the libraries
+ * @param toolsString The string with all the tools
+ * @param softwareString The string with all the software
+ * @param includeNames The boolean that determines whether or not to include names
+ * @returns The SVG with all the skills that were passed.
+ */
 export default function skillsWidget(
-    skillsString: string,
-    includeNames: boolean
+    languagesString?: string,
+    frameworksString?: string,
+    librariesString?: string,
+    toolsString?: string,
+    softwareString?: string,
+    includeNames?: boolean,
+
 ): string {
-    const skillList: Array<string> = skillsString.split(',')
 
-    if (skillList === undefined) {
-        return errorWidget('Skills', '-24%', 'Languages are undefined!', '-28%')
+    const BASE_HEIGHT = 125
+    const BASE_WIDTH = 812
+    const FIRST_ROW = 90 
+    const ROW = 114
+    const PAD = 60
+    
+    if (languagesString === undefined || languagesString === null) {
+        languagesString = 'undefined'
+    }
+    if (frameworksString === undefined || frameworksString === null) {
+        frameworksString = 'undefined'
+    }
+    if (librariesString === undefined || librariesString === null) {
+        librariesString = 'undefined'
+    }
+    if (toolsString === undefined || toolsString === null) {
+        toolsString = 'undefined'
+    }
+    if (softwareString === undefined || softwareString === null) {
+        softwareString = 'undefined'
     }
 
-    const width = 812
+    const languageList: string[] = languagesString.split(',')
+    const frameworkList: string[] = frameworksString.split(',')
+    const libraryList: string[] = librariesString.split(',')
+    const toolsList: string[] = toolsString.split(',')
+    const softwareList: string[] = softwareString.split(',')
+
+    const rowHeightLanguages = Math.round((languageList.length - 0.1) / 7) > 1 ? Math.round((languageList.length - 0.1) / 7) : 1
+    const languagesTitleHeight = FIRST_ROW 
+
+    const rowHeightFrameworks = Math.round((frameworkList.length - 0.1) / 7) > 1 ? Math.round((frameworkList.length - 0.1) / 7) : 1
+    const frameworkTitleHeight = languagesTitleHeight 
+    + ((languageList.length > 1 || languageList[0] !== 'undefined' ? 1  : 0) * PAD) 
```

Comment:
To add to the suggestion of an object array based approach, something like that would make this code also 10x better, as you wouldn't have to do weird lookups like checking if the first array element is equal to a string of 'undefined'

Reply:
Agreed, definitely something for a future version